aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+                           // guts::function_traits below extracts return_type and parameter_types from Registered,
+                           // which WrapFunction_ templates above use to declare their call methods.
+         class Redispatch, // The signature for the function we're redispatching to.  In most cases this is the same
+                           // as Registered, but for some ops (for example, ops where we append a dtype) it's useful
+                           // to redispatch to a function with a different signature.
+         Redispatch* F>    // The actual function we're redispatching to.
+struct WrapFunction final {
+  using type = WrapFunction_<policy,
+                             Redispatch,
+                             F,
+                             typename guts::function_traits<Registered>::return_type,
+                             typename guts::function_traits<Registered>::parameter_types>;
+};
+
+// Aside from CastPolicy::fp16 and CastPolicy::fp32, the guts of each ::call below need divergent code.
+// Some need to flip a dtype, some need to append a dtype, some need to swap in a new casted vector in
+// place of a TensorList, etc, so I declare specializations.
+// I could deduplicate these by having a single template with more complex (one-size-fits-all) type selection and
+// TensorList handling, but with specializations, each instantiation can compile with only the logic it needs.
+
+// CastPolicy::fp16
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp16, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    return (*F)(cached_cast(at::kHalf, args)...);
+  }
+};
+
+// CastPolicy::fp16_with_tensorlist
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp16_with_tensorlist, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    // The TensorList that receives vec will use vec's raw data pointer.
+    // Create vec out of line to ensure its data's lifetime lasts the full duration of (*F)().
+    std::vector<Tensor> vec = get_casted_vector(at::kHalf, args...);
+    return (*F)(cached_cast_with_vector(at::kHalf, vec, args)...);
+  }
+};
+
+// CastPolicy::fp32
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp32, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    return (*F)(cached_cast(at::kFloat, args)...);
+  }
+};
+
+// CastPolicy::fp32_set_opt_dtype
```

Comment:
Ooooof lol

Reply:
this sounds ominous but im not sure how to interpret it